LaunchServer+LibDesktop: Allow opening from mime type

demostanis opened the pull request:
This PR makes LaunchServer open files depending on the mime type based on preferences in `LaunchServer.ini` (which I have not written yet), and fallback to the old method by looking at the extension if the former fails.

Test case:
- Add `[MimeType]` section to `LaunchServer.ini`:
```
[MimeType]
image/png=/bin/ImageViewer
```
- Get some PNG file and rename its extension:
```
$ cp /res/graphics/buggie.png .
$ mv buggie.png buggie.someextensionthatobviouslydoesntexist
$ open buggie.someextensionthatobviouslydoesntexist
```
- ???
- Profit

PS: That's my first contribution to SerenityOS :^)

AtkinsSJ commented on Userland/Services/LaunchServer/Launcher.cpp:
```diff
@@ -156,6 +169,21 @@ Vector<String> Launcher::handlers_with_details_for_url(const URL& url)
+{
+    auto file_or_error = Core::File::open(path, Core::OpenMode::ReadOnly);
+    if (file_or_error.is_error()) {
+        return Optional<String> {};
```
```suggestion
        return {};
```

AtkinsSJ commented on Userland/Services/LaunchServer/Launcher.cpp:
```diff
@@ -312,10 +348,15 @@ bool Launcher::open_file_url(const URL& url)
+    auto extension = LexicalPath::extension(url.path()).to_lowercase();
+    auto mime_type = mime_type_for_file(url.path());
+
+    auto handlers = m_file_handlers;
+    auto mime_type_or_extension = extension;
+    if (mime_type.has_value()) {
+        handlers = m_mime_handlers;
+        mime_type_or_extension = mime_type.value();
+    }
```
Because `handlers` is a value not a reference, you're copying a HashTable when you initialise it, and copying another HashTable if there's a mime-type. You can avoid that by removing it, and then selecting the correct HashTable below. (See next comment)

```suggestion
    auto mime_type_or_extension = extension;
    if (mime_type.has_value()) {
        mime_type_or_extension = mime_type.value();
    }
```

AtkinsSJ commented on Userland/Services/LaunchServer/Launcher.cpp:
```diff
@@ -339,6 +380,6 @@ bool Launcher::open_file_url(const URL& url)
     additional_parameters.append(filepath);
 
-    return open_with_user_preferences(m_file_handlers, extension, additional_parameters, default_handler);
+    return open_with_user_preferences(handlers, mime_type_or_extension, additional_parameters, default_handler);
```
```suggestion
    return open_with_user_preferences(mime_type.has_value() ? m_mime_handlers : m_file_handlers, mime_type_or_extension, additional_parameters, default_handler);
```

AtkinsSJ reviewed (changes requested):
Welcome to Serenity! :^)

This is a cool feature, I do have a couple of comments on it, but overall the code looks good. :+1:

demostanis commented:
@AtkinsSJ: I incorporated your suggestions, and I also added a line to ensure we still fallback to extension-based opening of files if there are no handler for a mime type (even if we deduced the mime type).

demostanis commented:
For some reason git also committed the example changes in LaunchServer.ini although I had told it not to:
![image](https://user-images.githubusercontent.com/40673815/180492703-a6d45776-054d-4422-80b9-b31d75f627c0.png)
Git is a mystery to me...

kleinesfilmroellchen commented on Userland/Services/LaunchServer/Launcher.cpp:
```diff
@@ -83,6 +84,9 @@ void Launcher::load_handlers(String const& af_dir)
     Desktop::AppFile::for_each([&](auto af) {
         auto app_name = af->name();
         auto app_executable = af->executable();
+        HashTable<String> mime_types;
```
It looks like you forgot to pass this to an extended `Handler` constructor.

kleinesfilmroellchen commented on Userland/Services/LaunchServer/Launcher.cpp:
```diff
@@ -156,6 +169,21 @@ Vector<String> Launcher::handlers_with_details_for_url(const URL& url)
+    } else {
+        auto file = file_or_error.release_value();
+        // Read accounts for longest possible offset + signature we currently match against.
+        // Taken from Userland/Utilities/file.cpp
```
not necessary, this is our code, we can steal it however we want :^)

kleinesfilmroellchen reviewed (changes requested):
🥳🦬

kleinesfilmroellchen reviewed (approved):
Apart from the example, LGTM

demostanis commented:
I managed to remove the example

demostanis commented:
While I made those changes, I tried to understand what `[Launcher]` in `.af` files was for.
It seems like it has no effect, since everything is set in `LaunchServer.ini`. Or have my changes broke it?

demostanis commented:
I have investigated the above, it seems that the file manager can open files depending on what an `.af` file has in the `[FileType]` (or `[MimeType]` thanks to my PR) but that the Terminal can't.
That's gonna be a bug to fix in another PR...

demostanis commented:
Well that's because there is `m_handlers` and `m_{file,protocol,mime}_handlers`. `open` uses `m_handlers` while the file manager uses the others. Is there any reason they're split?

demostanis commented:
Fixed it. Should it belong to another PR?
Tell me if the commit wording isn't good or that my code sucks.

To test it:
- Remove an entry from `LaunchServer.ini`:
```
$ sed /png/d Base/home/anon/.config/LaunchServer.ini
```
- Try to open a file:
```
$ open /res/graphics/buggie.png
```
This also works for mime types, but you need to modify `.af` files:
```
$ echo MimeTypes=image/png >> Base/res/apps/ImageViewer.af
```
```
$ open buggie.someextensionthatobviouslydoesntexist
```

FireFox317 commented on Userland/Services/LaunchServer/Launcher.cpp:
```diff
@@ -227,14 +235,20 @@ Handler Launcher::get_handler_for_executable(Handler::Type handler_type, String
-    if (program_path.has_value())
-        return spawn(program_path.value(), arguments);
+    String executable = "";
+    if(for_each_handler(key, user_preferences, [&](auto const& handler) -> bool {
```
The `if(` looks incorrectly formatted, did you run `clang-format` on the code?

Btw, we have commit hooks that automatically run certain checks, including the formatting of the code, see: https://github.com/SerenityOS/serenity/blob/master/CONTRIBUTING.md#commit-hooks

demostanis replied:
Oops, I just installed the pre-commit hooks, and fixed formatting in the latest commit.

FireFox317 commented on Userland/Services/LaunchServer/Launcher.h:
```diff
@@ -50,12 +50,13 @@ class Launcher {
     HashMap<String, String> m_file_handlers;
     HashMap<String, String> m_mime_handlers;
 
+		bool has_mime_handlers(String const&);
```
Again, formatting here has gone wrong :)

kleinesfilmroellchen commented on Userland/Services/LaunchServer/Launcher.cpp:
```diff
@@ -156,6 +169,20 @@ Vector<String> Launcher::handlers_with_details_for_url(const URL& url)
+    } else {
+        auto file = file_or_error.release_value();
+        // Read accounts for longest possible offset + signature we currently match against.
+        auto bytes = file->read(0x9006);
```
No need to fix this, but we should probably compute that sniff size automatically... I think I just yakbait-ed myself.

kleinesfilmroellchen commented on Userland/Services/LaunchServer/Launcher.cpp:
```diff
@@ -227,14 +235,20 @@ Handler Launcher::get_handler_for_executable(Handler::Type handler_type, String
     return handler;
 }
 
-bool Launcher::open_with_user_preferences(HashMap<String, String> const& user_preferences, String const& key, Vector<String> const& arguments, String const& default_program)
+bool Launcher::open_with_user_preferences(HashMap<String, String>& user_preferences, String const& key, Vector<String> const& arguments, String const& default_program)
```
Why must the user preferences not be const? Is `for_each_handler` modifying them; logically speaking it should not?

demostanis replied:
That's right; I had made a change that would require it which I later removed, I fixed it back.

kleinesfilmroellchen commented on Userland/Services/LaunchServer/Launcher.cpp:
```diff
@@ -263,6 +277,7 @@ void Launcher::for_each_handler(String const& key, HashMap<String, String>& user
     auto user_default = user_preference.get("*");
     if (counted == 0 && user_default.has_value())
         f(get_handler_for_executable(Handler::Type::UserDefault, user_default.value()));
+    return counted;
```
Please leave a comment somewhere that explains what the return value of the function means as I find it unintuitive.

demostanis replied:
```
     // Return the number of times f() was called,
    // which can be used to know whether there were any handlers
```
That's used in `open_with_user_preferences` so that we don't open more than one app per file

kleinesfilmroellchen commented on Userland/Services/LaunchServer/Launcher.cpp:
```diff
@@ -303,19 +318,24 @@ void Launcher::for_each_handler_for_path(String const& path, Function<bool(Handl
-                return f(handler);
-            return false;
-        });
+        goto handle_extension;
```
I think this is entirely unnecessary; please avoid `goto` at almost all costs.

demostanis replied:
You're rightttt.

demostanis commented:
Looks like I committed the wrong stuff...

demostanis commented:
Looks like it's good. I fixed a bug where double clicking on commands in the Terminal wouldn't work.

AtkinsSJ commented on Userland/Services/LaunchServer/Launcher.cpp:
```diff
@@ -233,6 +241,16 @@ bool Launcher::open_with_user_preferences(HashMap<String, String> const& user_pr
+                return true;
+            }
+            return false;
+        }))
+        return spawn(executable, arguments);
```
Nit-pick: Wrapping this in `{}` would make it easier to understand. It's quite far from the `if`.

AtkinsSJ commented on Userland/Services/LaunchServer/Launcher.cpp:
```diff
@@ -303,19 +324,21 @@ void Launcher::for_each_handler_for_path(String const& path, Function<bool(Handl
+                    return f(handler);
+                return false;
+            }))
+            return;
```
Nit-pick: Again, wrap this line in `{}` please.

AtkinsSJ reviewed (changes requested):
Just a couple of nit-picks, then this is good to merge. :^)

demostanis commented:
Here :^)
